File: src/main/java/com/duxtinto/intellij/plugin/github/codereviews/ui/toolwindow/pullrequestlist/PullRequestList.java
package com.duxtinto.intellij.plugin.github.codereviews.ui.toolwindow.pullrequestlist;

import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.PullRequestEntity;
import com.intellij.util.ui.ListTableModel;

import javax.swing.*;
import java.util.List;

public interface PullRequestList {
    interface Model {
        ListTableModel<PullRequestEntity> getTableModel();
        void setPullRequests(List<PullRequestEntity> pullRequests);
    }
    interface View {
        void render(Model model);
        JComponent getContent();
    }
    interface Presenter {
        void displayPullRequests(List<PullRequestEntity> pullRequests);
        Model getModel();
        void setView(View view);
    }
}

File: src/main/kotlin/com/duxtinto/intellij/plugin/github/codereviews/ui/toolwindow/pullrequestlist/PullRequestList.kt
package com.duxtinto.intellij.plugin.github.codereviews.ui.toolwindow.pullrequestlist

import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.PullRequestEntity
import com.intellij.util.ui.ListTableModel
import javax.swing.JComponent
import javax.swing.event.ListSelectionListener
import java.awt.event.MouseListener as AwtMouseListener

interface PullRequestList {
    interface Model {
        val tableModel: ListTableModel<PullRequestEntity>
        fun setPullRequests(pullRequests: List<PullRequestEntity>)
    }

    interface View {
        val content: JComponent
        fun render(model: Model)

        interface Events {
            interface MouseListener : AwtMouseListener
            interface SelectionListener : ListSelectionListener
        }
    }

    interface Presenter {
        val model: Model
        fun displayPullRequests(pullRequests: List<PullRequestEntity>)
        fun setView(view: View)
    }
}
